src/descriptor/create_descriptor.rs

```diff
@@ -0,0 +1,410 @@
+/// elements from scriptsig into Vec<StackElement>
+fn verify_p2sh(
+    script_pubkey: &bitcoin::Script,
+    script_sig: &bitcoin::Script,
+) -> Result<(Script, Vec<StackElement>), Error>
+{
+    let (redeem_script, stack) = parse_scriptsig_top(script_sig)?;
+    let redeem_script = Script::from(redeem_script);
+    if redeem_script.to_p2sh() != *script_pubkey {
+        return Err(Error::BadDescriptor)
+    }
+    Ok((redeem_script, stack))
+}
+
+/// Figures out the the type of descriptor based on scriptpubkey, witness and scriptsig.
+/// Outputs a descriptor and witness_stack for all descriptors which can be directly fed into the
+/// interpreter. All script_sig and witness are translated into a single witness stack
+/// Vec<StackElement>
+/// 1) Pkh: Removes top element(pk) and validates pubkey hash, pushes rest of witness to
+/// witness_stack<StackElement> and outputs a Pkh descriptor
+/// 2) Wphk: translates witness to witness_stack<StackElement>, validates sig and pubkey hash
+/// and outputs a Wpkh descriptor
+/// 3) Wsh: pops witness script and checks wsh output hash, translates remaining witness elements
+/// to Vec<StackElement> and outputs a Wsh descriptor. Does not check miniscript inside the descriptor
+/// 4) Bare: translates script_sig to witness_stack<StackElement> and script_pubkey to miniscript
+/// 5) Sh: Checks redeem_script hash, translates remaining elements from script_sig to
+/// witness_stack<StackElement> and redeem script to miniscript. Does not check the miniscript
+/// 6) ShWpkh: Sh: Checks redeem_script hash, translates remaining elements from script_sig to
+/// witness_stack<StackElement> and validates Wpkh sig, pubkey.
+/// 7) ShWsh: Checks witness script hash, pops witness script and converts it to miniscript.
+/// translates the remaining witness to witness_stack<StackElement>
+pub fn witness_stack(
+    script_pubkey: &bitcoin::Script,
+    script_sig: &bitcoin::Script,
+    witness: Vec<Vec<u8>>,
+) -> Result<(Descriptor<PublicKey>, Vec<StackElement>), Error>
+{
+    if script_pubkey.is_p2pkh(){
+        verify_p2pkh(script_pubkey, script_sig, witness)
+    }
+    else if script_pubkey.is_v0_p2wpkh(){
+        let (pk, stack) = verify_wpkh(script_pubkey, script_sig, witness)?;
+        Ok((Descriptor::Wpkh(pk), stack))
+    }
+    else if script_pubkey.is_v0_p2wsh(){
+        let (ms, stack) = verify_wsh(script_pubkey, script_sig, witness)?;
+        Ok((Descriptor::Wsh(ms), stack))
+    }
+    else if script_pubkey.is_p2sh(){
+        let (redeem_script, stack) = verify_p2sh(script_pubkey, script_sig)?;
+        if redeem_script.is_v0_p2wpkh()
+        {
+            //ensures that scriptsig after popping redeem script contains 0 elements.
+            //Therefore while calling verify_wpkh, an argument of Script::new() is passed instead
+            //of script_sig. The redeem_script becomes the script_pubkey
+            if stack.len() != 0{
```

Comment:
nit: it's a bit more idiomatic to use `stack.is_empty()` rather than comparing `len` to 0. (And also for some non-`Vec` collections it's much more efficient).

This is something `clippy` will warn you about BTW ... if it's not too noisy (as I mentioned, it hasn't been compiling for me lately so I haven't used it) it's definitely worth running.